sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/ShardReadersPool.java

```diff
@@ -159,4 +216,76 @@ ShardRecordsIterator createShardIterator(SimplifiedKinesisClient kinesis,
     return new ShardRecordsIterator(checkpoint, kinesis);
   }
 
+  /**
+   * Waits until all records read from given shardRecordsIterator are taken from
+   * {@link #recordsQueue} and acked.
+   * Uses {@link #numberOfRecordsInAQueueByShard} map to track the amount of remaining events.
+   */
+  private void waitUntilAllShardRecordsRead(ShardRecordsIterator shardRecordsIterator)
+      throws InterruptedException {
+    while (!allShardRecordsRead(shardRecordsIterator)) {
+      Thread.sleep(TimeUnit.SECONDS.toMillis(1));
+    }
+  }
+
+  private boolean allShardRecordsRead(final ShardRecordsIterator shardRecordsIterator) {
+    return numberOfRecordsInAQueueByShard.get(shardRecordsIterator.getShardId()).get() == 0;
+  }
+
+  /**
+   * <p>
+   * Tries to find successors of a given shard and start reading them. Each closed shard can have
+   * 0, 1 or 2 successors
+   * <ul>
+   *   <li>0 successors - when shard was merged with another shard and this one is considered
+   *   adjacent by merge operation</li>
+   *   <li>1 successor - when shard was merged with another shard and this one is considered a
+   *   parent by merge operation</li>
+   *   <li>2 successors - when shard was split into two shards</li>
+   * </ul>
+   * </p>
+   * <p>
+   * Once shard successors are established, the transition to reading new shards can begin.
+   * During this operation, the immutable {@link ShardReadersPool#shardIteratorsMap}
+   * is replaced with a new one holding references to {@link ShardRecordsIterator} instances for
+   * open shards only. Potentially there might be more shard iterators closing at the same time so
+   * {@link ShardReadersPool#shardIteratorsMap} is updated in a loop using CAS pattern to keep all
+   * the updates. Then, the counter for already closed shard is removed from
+   * {@link ShardReadersPool#numberOfRecordsInAQueueByShard} map.
+   * </p>
+   * <p>
+   * Finally when update is finished, new threads are spawned for reading the successive shards.
+   * The thread that handled reading from already closed shard can finally complete.
+   * </p>
+   */
+  private void readFromSuccessiveShards(final ShardRecordsIterator closedShardIterator)
+      throws TransientKinesisException {
+    List<ShardRecordsIterator> successiveShardRecordIterators = closedShardIterator
+        .findSuccessiveShardRecordIterators();
+
+    ImmutableMap<String, ShardRecordsIterator> current;
+    ImmutableMap<String, ShardRecordsIterator> updated;
+    do {
```

Comment:
This seems to beg for a simple `synchronized` instead of a CAS loop. I really don't see why we can't just have `synchronized` regions around manipulations with `shardIteratorsMap` and `numberOfRecords...` - I think with the simplified handling of the queue, this should be even simpler now?

One more issue in the current code: while this loop runs, the value of `successiveShardRecordIterators` may get out of date (e.g. some other shards might have been closed and ran this loop). This is another argument in favor of making the entire shard rebalancing operation a big `synchronized` block, and then figure out how to ensure that it doesn't block enqueueing or reading records - seems it should be not too hard.

Reply:
`successiveShardRecordIterators` value is computed for given shard and given shard will always have the same successors no matter what happens to other shards. This is safe.

I just don't see a reason for using `synchronized` here or anywhere else anymore. `recordsQueue` is just a concurrent collection, there's no atomic reference to it. `shardIteratorsMap` is an atomic reference to an immutable map and `numberOfRecordsInAQueueByShard` is a concurrent map, both are safe on their own. Both are changed in this place only, during a highly infrequent operation and do not have to be modified atomically.

In my opinion using `synchronized` on all accesses to `shardIteratorsMap` and `numberOfRecordsInAQueueByShard` will just clutter the code.